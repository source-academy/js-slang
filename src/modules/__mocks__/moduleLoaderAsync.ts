import { Context } from '../../types'

export const memoizedGetModuleDocsAsync = jest.fn().mockResolvedValue({
  foo: 'foo',
  bar: 'bar'
})

export const memoizedGetModuleBundleAsync = jest.fn().mockResolvedValue(
  `require => ({
    foo: () => 'foo',
    bar: () => 'bar',
  })`
)

export const memoizedGetModuleManifestAsync = jest.fn().mockResolvedValue({
  one_module: { tabs: [] },
  other_module: { tabs: [] },
  another_module: { tabs: [] }
})

export function loadModuleBundleAsync() {
  return Promise.resolve({
    foo: () => 'foo',
    bar: () => 'bar'
  })
}

export function loadModuleTabsAsync(_name: string) {
  return Promise.resolve([])
}

export async function initModuleContextAsync(
  moduleName: string,
  context: Context,
  loadTabs: boolean
) {
  // Load the module's tabs
  if (!(moduleName in context.moduleContexts)) {
    context.moduleContexts[moduleName] = {
      state: null,
      tabs: loadTabs ? await loadModuleTabsAsync(moduleName) : null
    }
  } else if (context.moduleContexts[moduleName].tabs === null && loadTabs) {
    context.moduleContexts[moduleName].tabs = await loadModuleTabsAsync(moduleName)
  }
}
