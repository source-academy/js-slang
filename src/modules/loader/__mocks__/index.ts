import type { Context } from '../../..'
import {
  unknownDocs,
  type FunctionDocumentation,
  type ModuleFunctions,
  type VariableDocumentation
} from '../../moduleTypes'

export const memoizedGetModuleDocsAsync = jest.fn((module: string) => {
  const barDocs: FunctionDocumentation = {
    kind: 'function',
    retType: 'void',
    params: [['a', 'number']],
    description: 'bar'
  }

  const fooDocs: VariableDocumentation = {
    kind: 'variable',
    type: 'string',
    description: 'foo'
  }

  return Promise.resolve(
    module === 'another_module'
      ? {
          bar: barDocs,
          foo: fooDocs
        }
      : {
          bar: barDocs,
          foo: fooDocs,
          default: unknownDocs
        }
  )
})

export const memoizedGetModuleManifestAsync = jest.fn().mockResolvedValue({
  one_module: { tabs: [] },
  another_module: { tabs: [] },
  other_module: { tabs: [] }
})

export const loadModuleBundleAsync = jest.fn((name: string) => {
  const baseModule = {
    foo: () => 'foo',
    bar: () => 'bar'
  }

  return Promise.resolve(
    name === 'another_module' ? baseModule : { ...baseModule, default: () => 'def' }
  )
})

export const loadModuleTabsAsync = jest.fn().mockResolvedValue([])

async function initModuleContextAsync(moduleName: string, context: Context, loadTabs: boolean) {
  // Load the module's tabs
  if (!(moduleName in context.moduleContexts)) {
    context.moduleContexts[moduleName] = {
      state: null,
      tabs: loadTabs ? await loadModuleTabsAsync(moduleName) : null
    }
  } else if (context.moduleContexts[moduleName].tabs === null && loadTabs) {
    context.moduleContexts[moduleName].tabs = await loadModuleTabsAsync(moduleName)
  }
}

export default async function loadSourceModules(
  sourceModulesToImport: Set<string>,
  context: Context,
  loadTabs: boolean
) {
  const loadedModules = await Promise.all(
    [...sourceModulesToImport].map(async moduleName => {
      await initModuleContextAsync(moduleName, context, loadTabs)
      const bundle = await loadModuleBundleAsync(moduleName)
      return [moduleName, bundle] as [string, ModuleFunctions]
    })
  )
  context.nativeStorage.loadedModules = Object.fromEntries(loadedModules)
}
