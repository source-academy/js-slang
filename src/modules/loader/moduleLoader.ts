// Export for compatibility with frontend
export { setModulesStaticURL } from '.'
