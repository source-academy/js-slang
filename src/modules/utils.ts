/**
 * For tabs that are export default declarations, we need to remove
 * the `export default` bit from the front before they can be loaded
 * by js-slang
 */
export function evalRawTab(text: string) {
  if (text.startsWith('export default')) {
    text = text.substring(14)
  }

  return eval(text)
}
