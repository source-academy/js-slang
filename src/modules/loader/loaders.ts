import type { Context, Node } from '../../types'
import { timeoutPromise } from '../../utils/misc'
import { ModuleConnectionError, ModuleInternalError } from '../errors'
import type {
  ModuleBundle,
  ModuleDocumentation,
  ModuleFunctions,
  ModuleManifest
} from '../moduleTypes'
import { getRequireProvider } from './requireProvider'

/** Default modules static url. Exported for testing. */
export let MODULES_STATIC_URL = 'https://source-academy.github.io/modules'

export function setModulesStaticURL(url: string) {
  MODULES_STATIC_URL = url

  // Changing the module backend should clear these
  memoizedGetModuleDocsAsync.cache.clear()
  memoizedGetModuleManifestAsync.reset()
}

function wrapImporter<T>(func: (p: string) => Promise<T>) {
  /*
    Browsers natively support esm's import() but Jest and Node do not. So we need
    to change which import function we use based on the environment.

    For the browser, we use the function constructor to hide the import calls from
    webpack so that webpack doesn't try to compile them away.

    Browsers automatically cache import() calls, so we add a query parameter with the
    current time to always invalidate the cache and handle the memoization ourselves
  */
  return async (p: string): Promise<T> => {
    try {
      const result = await timeoutPromise(func(p), 10000)
      return result
    } catch (error) {
      // Before calling this function, the import analyzer should've been used to make sure
      // that the module being imported already exists, so the following errors should
      // be thrown only if the modules server is unreachable
      if (
        // In the browser, import statements should throw TypeError
        (typeof window !== 'undefined' && error instanceof TypeError) ||
        // In Node a different error is thrown with the given code instead
        error.code === 'MODULE_NOT_FOUND' ||
        // Thrown specifically by jest
        error.code === 'ENOENT'
      ) {
        throw new ModuleConnectionError()
      }
      throw error
    }
  }
}

// Exported for testing
export const docsImporter = wrapImporter<{ default: any }>(async p => {
  // TODO: USe import attributes when they become supported
  // Import Assertions and Attributes are not widely supported by all
  // browsers yet, so we use fetch in the meantime
  const resp = await fetch(p)
  if (resp.status !== 200 && resp.status !== 304) {
    throw new ModuleConnectionError()
  }

  const result = await resp.json()
  return { default: result }
})

// lodash's memoize function memoizes on errors. This is undesirable,
// so we have our own custom memoization that won't memoize on errors
function getManifestImporter() {
  let manifest: ModuleManifest | null = null

  async function func() {
    if (manifest !== null) {
      return manifest
    }

    ;({ default: manifest } = await docsImporter(`${MODULES_STATIC_URL}/modules.json`))

    return manifest!
  }

  func.reset = () => {
    manifest = null
  }

  return func
}

function getMemoizedDocsImporter() {
  const docs = new Map<string, ModuleDocumentation>()

  async function func(moduleName: string, throwOnError: true): Promise<ModuleDocumentation>
  async function func(moduleName: string, throwOnError?: false): Promise<ModuleDocumentation | null>
  async function func(moduleName: string, throwOnError?: boolean) {
    if (docs.has(moduleName)) {
      return docs.get(moduleName)!
    }

    try {
      const { default: loadedDocs } = await docsImporter(
        `${MODULES_STATIC_URL}/jsons/${moduleName}.json`
      )
      docs.set(moduleName, loadedDocs)
      return loadedDocs
    } catch (error) {
      if (throwOnError) throw error
      console.warn(`Failed to load documentation for ${moduleName}:`, error)
      return null
    }
  }

  func.cache = docs
  return func
}

export const memoizedGetModuleManifestAsync = getManifestImporter()
export const memoizedGetModuleDocsAsync = getMemoizedDocsImporter()

const bundleAndTabImporter = wrapImporter<{ default: ModuleBundle }>(
  typeof window !== 'undefined' && process.env.NODE_ENV !== 'test'
    ? (new Function('path', 'return import(`${path}?q=${Date.now()}`)') as any)
    : p => Promise.resolve(require(p))
)

export async function loadModuleBundleAsync(
  moduleName: string,
  context: Context,
  node?: Node
): Promise<ModuleFunctions> {
  const { default: result } = await bundleAndTabImporter(
    `${MODULES_STATIC_URL}/bundles/${moduleName}.js`
  )
  try {
    const loadedModule = result(getRequireProvider(context))
    return Object.entries(loadedModule).reduce((res, [name, value]) => {
      if (typeof value === 'function') {
        const repr = `function ${name} {\n\t[Function from ${moduleName}\n\tImplementation hidden]\n}`
        value[Symbol.toStringTag] = () => repr
        value.toString = () => repr
      }
      return {
        ...res,
        [name]: value
      }
    }, {})
  } catch (error) {
    throw new ModuleInternalError(moduleName, error, node)
  }
}

export async function loadModuleTabsAsync(moduleName: string) {
  const manifest = await memoizedGetModuleManifestAsync()
  const moduleInfo = manifest[moduleName]

  return Promise.all(
    moduleInfo.tabs.map(async tabName => {
      const { default: result } = await bundleAndTabImporter(
        `${MODULES_STATIC_URL}/tabs/${tabName}.js`
      )
      return result
    })
  )
}
