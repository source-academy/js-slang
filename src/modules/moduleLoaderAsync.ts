import type { Node } from 'estree'
import { memoize } from 'lodash'

import type { Context } from '..'
import { PromiseTimeoutError, timeoutPromise } from '../utils/misc'
import { wrapSourceModule } from '../utils/operators'
import { ModuleConnectionError, ModuleInternalError, ModuleNotFoundError } from './errors'
import { MODULES_STATIC_URL } from './moduleLoader'
import type { ModuleBundle, ModuleDocumentation, ModuleManifest } from './moduleTypes'
import { getRequireProvider } from './requireProvider'
import { evalRawTab } from './utils'

export function httpGetAsync(path: string, type: 'json'): Promise<object>
export function httpGetAsync(path: string, type: 'text'): Promise<string>
export async function httpGetAsync(path: string, type: 'json' | 'text') {
  try {
    const resp = await timeoutPromise(
      fetch(path, {
        method: 'GET'
      }),
      10000
    )

    if (resp.status !== 200 && resp.status !== 304) {
      throw new ModuleConnectionError()
    }

    const promise = type === 'text' ? resp.text() : resp.json()
    return timeoutPromise(promise, 10000)
  } catch (error) {
    if (error instanceof TypeError || error instanceof PromiseTimeoutError) {
      throw new ModuleConnectionError()
    }
    if (!(error instanceof ModuleConnectionError)) throw new ModuleInternalError(path, error)
    throw error
  }
}

/**
 * Send a HTTP GET request to the modules endpoint to retrieve the manifest
 * @return Modules
 */
export const memoizedGetModuleManifestAsync = memoize(getModuleManifestAsync)
function getModuleManifestAsync(): Promise<ModuleManifest> {
  return httpGetAsync(`${MODULES_STATIC_URL}/modules.json`, 'json') as Promise<ModuleManifest>
}

async function checkModuleExists(moduleName: string, node?: Node) {
  const modules = await memoizedGetModuleManifestAsync()
  // Check if the module exists
  if (!(moduleName in modules)) throw new ModuleNotFoundError(moduleName, node)

  return modules[moduleName]
}

export const memoizedGetModuleBundleAsync = memoize(getModuleBundleAsync)
async function getModuleBundleAsync(moduleName: string): Promise<string> {
  return httpGetAsync(`${MODULES_STATIC_URL}/bundles/${moduleName}.js`, 'text')
}

export const memoizedGetModuleTabAsync = memoize(getModuleTabAsync)
function getModuleTabAsync(tabName: string): Promise<string> {
  return httpGetAsync(`${MODULES_STATIC_URL}/tabs/${tabName}.js`, 'text')
}

export const memoizedGetModuleDocsAsync = memoize(getModuleDocsAsync)
async function getModuleDocsAsync(moduleName: string): Promise<ModuleDocumentation | null> {
  try {
    const result = await httpGetAsync(`${MODULES_STATIC_URL}/jsons/${moduleName}.json`, 'json')
    return result as ModuleDocumentation
  } catch (error) {
    console.warn(`Failed to load documentation for ${moduleName}:`, error)
    return null
  }
}

export async function loadModuleTabsAsync(moduleName: string, node?: Node) {
  const moduleInfo = await checkModuleExists(moduleName, node)

  // Load the tabs for the current module
  return Promise.all(
    moduleInfo.tabs.map(async path => {
      const rawTabFile = await memoizedGetModuleTabAsync(path)
      try {
        return evalRawTab(rawTabFile)
      } catch (error) {
        // console.error('tab error:', error);
        throw new ModuleInternalError(path, error, node)
      }
    })
  )
}

export async function loadModuleBundleAsync(
  moduleName: string,
  context: Context,
  wrapModule: boolean,
  node?: Node
) {
  // await checkModuleExists(moduleName, node)
  const moduleText = await memoizedGetModuleBundleAsync(moduleName)
  try {
    const moduleBundle: ModuleBundle = eval(moduleText)

    if (wrapModule) return wrapSourceModule(moduleName, moduleBundle, getRequireProvider(context))
    return moduleBundle(getRequireProvider(context))
  } catch (error) {
    // console.error("bundle error: ", error)
    throw new ModuleInternalError(moduleName, error, node)
  }
}

/**
 * Initialize module contexts and add UI tabs needed for modules to program context
 */
export async function initModuleContextAsync(
  moduleName: string,
  context: Context,
  loadTabs: boolean
) {
  // Load the module's tabs
  if (!(moduleName in context.moduleContexts)) {
    context.moduleContexts[moduleName] = {
      state: null,
      tabs: loadTabs ? await loadModuleTabsAsync(moduleName) : null
    }
  } else if (context.moduleContexts[moduleName].tabs === null && loadTabs) {
    context.moduleContexts[moduleName].tabs = await loadModuleTabsAsync(moduleName)
  }
}
