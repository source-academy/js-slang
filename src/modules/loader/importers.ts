import { timeoutPromise } from '../../utils/misc'
import { ModuleConnectionError } from '../errors'
import { memoizedGetModuleDocsAsync, memoizedGetModuleManifestAsync } from './loaders'

/** Default modules static url. Exported for testing. */
export let MODULES_STATIC_URL = 'https://source-academy.github.io/modules'

export function setModulesStaticURL(url: string) {
  MODULES_STATIC_URL = url

  // Changing the module backend should clear these
  memoizedGetModuleDocsAsync.cache.clear()
  memoizedGetModuleManifestAsync.reset()
}

function wrapImporter<T extends (p: string, signal?: AbortSignal) => Promise<any>>(func: T) {
  return async (p: string, signal?: AbortSignal): Promise<Awaited<ReturnType<T>>> => {
    try {
      const result = await timeoutPromise(func(p, signal), 10000)
      return result
    } catch (error) {
      console.error('The error is', error)
      // Before calling this function, the import analyzer should've been used to make sure
      // that the module being imported already exists, so the following errors should
      // be thrown only if the modules server is unreachable
      if (
        // In the browser, import statements should throw TypeError
        (typeof window !== 'undefined' && error instanceof TypeError) ||
        // In Node a different error is thrown with the given code instead
        error.code === 'MODULE_NOT_FOUND' ||
        // Thrown specifically by vitest
        (process.env.NODE_ENV === 'test' && error.code === 'ERR_UNSUPPORTED_ESM_URL_SCHEME')
      ) {
        throw new ModuleConnectionError()
      }
      throw error
    }
  }
}

// Exported for testing
export const docsImporter = wrapImporter(async (p, signal) => {
  // TODO: Use import attributes when they become supported
  // Import Assertions and Attributes are not widely supported by all
  // browsers yet, so we use fetch in the meantime
  const resp = await fetch(p, { signal })
  if (resp.status !== 200 && resp.status !== 304) {
    throw new ModuleConnectionError()
  }

  const result = await resp.json()
  return { default: result }
})

function getBundleAndTabsImporter(): (p: string) => Promise<any> {
  if (process.env.NODE_ENV === 'test') return p => import(p)
  if (typeof window !== 'undefined')
    return new Function('path', 'return import(`${path}?q=${Date.now()}`)') as any

  // eslint-disable-next-line @typescript-eslint/no-require-imports
  return p => Promise.resolve(require(p))
}

/*
  Browsers natively support esm's import() but Jest and Node do not. So we need
  to change which import function we use based on the environment.

  For the browser, we use the function constructor to hide the import calls from
  webpack so that webpack doesn't try to compile them away.
*/
export const bundleAndTabImporter = wrapImporter(getBundleAndTabsImporter())
