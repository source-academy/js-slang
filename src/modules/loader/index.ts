import type { Context } from '../../types'
import type { ModuleFunctions } from '../moduleTypes'
import { loadModuleBundleAsync, loadModuleTabsAsync } from './loaders'

/**
 * Initialize module contexts and add UI tabs needed for modules to program context
 */
async function initModuleContextAsync(moduleName: string, context: Context, loadTabs: boolean) {
  // Load the module's tabs
  if (!(moduleName in context.moduleContexts)) {
    context.moduleContexts[moduleName] = {
      state: null,
      tabs: loadTabs ? await loadModuleTabsAsync(moduleName) : null
    }
  } else if (context.moduleContexts[moduleName].tabs === null && loadTabs) {
    context.moduleContexts[moduleName].tabs = await loadModuleTabsAsync(moduleName)
  }
}

/**
 * With the given set of Source Modules to Import, load all of the bundles and
 * tabs (if `loadTabs` is true) and populate the `context.nativeStorage.loadedModules`
 * property.
 */
export default async function loadSourceModules(
  sourceModulesToImport: Set<string>,
  context: Context,
  loadTabs: boolean
) {
  const loadedModules = await Promise.all(
    [...sourceModulesToImport].map(async moduleName => {
      await initModuleContextAsync(moduleName, context, loadTabs)
      const bundle = await loadModuleBundleAsync(moduleName, context)
      return [moduleName, bundle] as [string, ModuleFunctions]
    })
  )

  const loadedObj = Object.fromEntries(loadedModules)
  context.nativeStorage.loadedModules = loadedObj

  return loadedObj
}

export {
  setModulesStaticURL,
  MODULES_STATIC_URL,
  memoizedGetModuleDocsAsync,
  memoizedGetModuleManifestAsync
} from './loaders'
