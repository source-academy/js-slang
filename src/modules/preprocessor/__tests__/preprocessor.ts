import type { Program } from 'estree'
import type { MockedFunction } from 'jest-mock'

import { parseError } from '../../..'
import { mockContext } from '../../../mocks/context'
import { Chapter } from '../../../types'
import { memoizedGetModuleDocsAsync } from '../../loader/moduleLoaderAsync'
import preprocessFileImports from '..'
import { parseCodeError, sanitizeAST } from './utils'
import { parse } from '../../../parser/parser'
import {
  accessExportFunctionName,
  defaultExportLookupName
} from '../../../stdlib/localImport.prelude'

jest.mock('../../loader/moduleLoaderAsync')

describe('preprocessFileImports', () => {
  let actualContext = mockContext(Chapter.LIBRARY_PARSER)
  let expectedContext = mockContext(Chapter.LIBRARY_PARSER)

  beforeEach(() => {
    actualContext = mockContext(Chapter.LIBRARY_PARSER)
    expectedContext = mockContext(Chapter.LIBRARY_PARSER)
  })

  const assertASTsAreEquivalent = (
    actualProgram: Program | undefined,
    expectedCode: string,
    log: boolean = false
  ): void => {
    if (!actualProgram) {
      throw new Error('Actual program should not be undefined!')
    }

    const expectedProgram = parse(expectedCode, expectedContext)
    if (!expectedProgram) {
      throw parseCodeError
    }

    expect(sanitizeAST(actualProgram)).toMatchObject(sanitizeAST(expectedProgram))
  }

  it('returns undefined & adds ModuleNotFoundError to context if the entrypoint file does not exist', async () => {
    const files: Record<string, string> = {
      '/a.js': '1 + 2;'
    }
    const actualProgram = await preprocessFileImports(files, '/non-existent-file.js', actualContext)
    expect(actualProgram).toBeUndefined()
    expect(parseError(actualContext.errors)).toMatchInlineSnapshot(
      `"Module '/non-existent-file.js' not found."`
    )
  })

  it('returns undefined & adds ModuleNotFoundError to context if an imported file does not exist', async () => {
    const files: Record<string, string> = {
      '/a.js': `import { x } from './non-existent-file.js';`
    }
    const actualProgram = await preprocessFileImports(files, '/a.js', actualContext)
    expect(actualProgram).toBeUndefined()
    expect(parseError(actualContext.errors)).toMatchInlineSnapshot(
      `"Line 1: Module '/non-existent-file.js' not found."`
    )
  })

  it('returns the same AST if the entrypoint file does not contain import/export statements', async () => {
    const files: Record<string, string> = {
      '/a.js': `
        function square(x) {
          return x * x;
        }
        square(5);
      `
    }
    const expectedCode = files['/a.js']
    const actualProgram = await preprocessFileImports(files, '/a.js', actualContext)
    assertASTsAreEquivalent(actualProgram, expectedCode)
  })

  it('removes all export-related AST nodes', async () => {
    const files: Record<string, string> = {
      '/a.js': `
        export const x = 42;
        export let y = 53;
        export function square(x) {
          return x * x;
        }
        export const id = x => x;
        export default function cube(x) {
          return x * x * x;
        }
      `
    }
    const expectedCode = `
      const x = 42;
      let y = 53;
      function square(x) {
       return x * x;
      }
      const id = x => x;
      function cube(x) {
       return x * x * x;
      }
    `
    const actualProgram = await preprocessFileImports(files, '/a.js', actualContext)
    assertASTsAreEquivalent(actualProgram, expectedCode)
  })

  it('ignores Source module imports & removes all non-Source module import-related AST nodes in the preprocessed program', async () => {
    const docsMocked = memoizedGetModuleDocsAsync as MockedFunction<
      typeof memoizedGetModuleDocsAsync
    >
    docsMocked.mockResolvedValueOnce({
      default: {} as any,
      a: {} as any,
      b: {} as any,
      c: {} as any
    })

    const files: Record<string, string> = {
      '/a.js': `
        import d, { a, b, c } from "one_module";
        import w, { x, y, z } from "./not-source-module.js";
      `,
      '/not-source-module.js': `
        export const x = 1;
        export const y = 2;
        export const z = 3;
        export default function square(x) {
          return x * x;
        }
      `
    }
    const expectedCode = `
      import d, { a, b, c } from "one_module";

      function __$not$$dash$$source$$dash$$module$$dot$$js__() {
        const x = 1;
        const y = 2;
        const z = 3;
        function square(x) {
          return x * x;
        }

        return pair(square, list(pair("x", x), pair("y", y), pair("z", z)));
      }

      const ___$not$$dash$$source$$dash$$module$$dot$$js___ = __$not$$dash$$source$$dash$$module$$dot$$js__();

      const w = ${accessExportFunctionName}(___$not$$dash$$source$$dash$$module$$dot$$js___, "${defaultExportLookupName}");
      const x = ${accessExportFunctionName}(___$not$$dash$$source$$dash$$module$$dot$$js___, "x");
      const y = ${accessExportFunctionName}(___$not$$dash$$source$$dash$$module$$dot$$js___, "y");
      const z = ${accessExportFunctionName}(___$not$$dash$$source$$dash$$module$$dot$$js___, "z");
    `
    const actualProgram = await preprocessFileImports(files, '/a.js', actualContext, {
      importOptions: {
        allowUndefinedImports: true
      },
      shouldAddFileName: true
    })
    assertASTsAreEquivalent(actualProgram, expectedCode, true)
  })

  it('collates Source module imports at the start of the top-level environment of the preprocessed program', async () => {
    const docsMocked = memoizedGetModuleDocsAsync as MockedFunction<
      typeof memoizedGetModuleDocsAsync
    >
    docsMocked.mockResolvedValue({
      f: {} as any,
      g: {} as any,
      h: {} as any,
      w: {} as any,
      x: {} as any,
      y: {} as any,
      z: {} as any
    })
    const files: Record<string, string> = {
      '/a.js': `
        import { b } from "./b.js";
        import { w, x } from "one_module";
        import { f, g } from "other_module";

        b;
      `,
      '/b.js': `
        import { square } from "./c.js";
        import { x, y } from "one_module";
        import { h } from "another_module";

        export const b = square(5);
      `,
      '/c.js': `
        import { x, y, z } from "one_module";

        export const square = x => x * x;
      `
    }
    const expectedCode = `
      import { w, x, y, z } from "one_module";
      import { f, g } from "other_module";
      import { h } from "another_module";

      function __$b$$dot$$js__(___$c$$dot$$js___) {
        const square = ${accessExportFunctionName}(___$c$$dot$$js___, "square");

        const b = square(5);

        return pair(null, list(pair("b", b)));
      }

      function __$c$$dot$$js__() {
        const square = x => x * x;

        return pair(null, list(pair("square", square)));
      }

      const ___$c$$dot$$js___ = __$c$$dot$$js__();
      const ___$b$$dot$$js___ = __$b$$dot$$js__(___$c$$dot$$js___);

      const b = ${accessExportFunctionName}(___$b$$dot$$js___, "b");

      b;
    `
    const actualProgram = await preprocessFileImports(files, '/a.js', actualContext, {
      importOptions: {
        allowUndefinedImports: true
      },
      shouldAddFileName: true
    })
    assertASTsAreEquivalent(actualProgram, expectedCode)
  })

  it('returns CircularImportError if there are circular imports', async () => {
    const files: Record<string, string> = {
      '/a.js': `
        import { b } from "./b.js";

        export const a = 1;
      `,
      '/b.js': `
        import { c } from "./c.js";

        export const b = 2;
      `,
      '/c.js': `
        import { a } from "./a.js";

        export const c = 3;
      `
    }
    await preprocessFileImports(files, '/a.js', actualContext, {
      shouldAddFileName: true
    })
    expect(parseError(actualContext.errors)).toMatchInlineSnapshot(
      `"Circular import detected: '/a.js' -> '/b.js' -> '/c.js' -> '/a.js'."`
    )
  })

  it('returns CircularImportError if there are circular imports - verbose', async () => {
    const files: Record<string, string> = {
      '/a.js': `
        import { b } from "./b.js";

        export const a = 1;
      `,
      '/b.js': `
        import { c } from "./c.js";

        export const b = 2;
      `,
      '/c.js': `
        import { a } from "./a.js";

        export const c = 3;
      `
    }
    await preprocessFileImports(files, '/a.js', actualContext)
    expect(parseError(actualContext.errors, true)).toMatchInlineSnapshot(`
      "Circular import detected: '/a.js' -> '/b.js' -> '/c.js' -> '/a.js'.
      Break the circular import cycle by removing imports from any of the offending files.
      "
    `)
  })

  it('returns CircularImportError if there are self-imports', async () => {
    const files: Record<string, string> = {
      '/a.js': `
        import { y } from "./a.js";
        const x = 1;
        export { x as y };
      `
    }
    await preprocessFileImports(files, '/a.js', actualContext)
    expect(parseError(actualContext.errors)).toMatchInlineSnapshot(
      `"Circular import detected: '/a.js' -> '/a.js'."`
    )
  })

  it('returns CircularImportError if there are self-imports - verbose', async () => {
    const files: Record<string, string> = {
      '/a.js': `
        import { y } from "./a.js";
        const x = 1;
        export { x as y };
      `
    }
    await preprocessFileImports(files, '/a.js', actualContext)
    expect(parseError(actualContext.errors, true)).toMatchInlineSnapshot(`
      "Circular import detected: '/a.js' -> '/a.js'.
      Break the circular import cycle by removing imports from any of the offending files.
      "
    `)
  })

  it('returns a preprocessed program with all imports', async () => {
    const files: Record<string, string> = {
      '/a.js': `
        import { a as x, b as y } from "./b.js";

        x + y;
      `,
      '/b.js': `
        import y, { square } from "./c.js";

        const a = square(y);
        const b = 3;
        export { a, b };
      `,
      '/c.js': `
        import { mysteryFunction } from "./d.js";

        const x = mysteryFunction(5);
        export function square(x) {
          return x * x;
        }
        export default x;
      `,
      '/d.js': `
        const addTwo = x => x + 2;
        export { addTwo as mysteryFunction };
      `
    }

    const expectedCode = `
      function __$b$$dot$$js__(___$c$$dot$$js___) {
        const y = ${accessExportFunctionName}(___$c$$dot$$js___, "${defaultExportLookupName}");
        const square = ${accessExportFunctionName}(___$c$$dot$$js___, "square");

        const a = square(y);
        const b = 3;

        return pair(null, list(pair("a", a), pair("b", b)));
      }

      function __$c$$dot$$js__(___$d$$dot$$js___) {
        const mysteryFunction = ${accessExportFunctionName}(___$d$$dot$$js___, "mysteryFunction");

        const x = mysteryFunction(5);
        function square(x) {
          return x * x;
        }

        return pair(x, list(pair("square", square)));
      }

      function __$d$$dot$$js__() {
        const addTwo = x => x + 2;

        return pair(null, list(pair("mysteryFunction", addTwo)));
      }

      const ___$d$$dot$$js___ = __$d$$dot$$js__();
      const ___$c$$dot$$js___ = __$c$$dot$$js__(___$d$$dot$$js___);
      const ___$b$$dot$$js___ = __$b$$dot$$js__(___$c$$dot$$js___);

      const x = ${accessExportFunctionName}(___$b$$dot$$js___, "a");
      const y = ${accessExportFunctionName}(___$b$$dot$$js___, "b");

      x + y;
    `
    const actualProgram = await preprocessFileImports(files, '/a.js', actualContext, {
      importOptions: {
        allowUndefinedImports: true
      },
      shouldAddFileName: true
    })

    assertASTsAreEquivalent(actualProgram, expectedCode)
  })
})
