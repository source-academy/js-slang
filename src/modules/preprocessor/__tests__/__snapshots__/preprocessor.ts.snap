// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preprocessFileImports collates Source module imports at the start of the top-level environment of the preprocessed program 1`] = `
"import {x, y, z, w} from \\"one_module\\";
import {h} from \\"another_module\\";
import {f, g} from \\"other_module\\";
import {x, y, z, w} from \\"one_module\\";
import {h} from \\"another_module\\";
import {f, g} from \\"other_module\\";
function __$c$$dot$$js__() {
  const square = x => x * x;
  return pair(null, list(pair(\\"square\\", square)));
}
function __$b$$dot$$js__(___$c$$dot$$js___) {
  const square = __access_export__(___$c$$dot$$js___, \\"square\\");
  const b = square(5);
  return pair(null, list(pair(\\"b\\", b)));
}
const ___$c$$dot$$js___ = __$c$$dot$$js__();
const ___$b$$dot$$js___ = __$b$$dot$$js__(___$c$$dot$$js___);
const b = __access_export__(___$b$$dot$$js___, \\"b\\");
b;
"
`;

exports[`preprocessFileImports ignores Source module imports & removes all non-Source module import-related AST nodes in the preprocessed program 1`] = `
"import {a, b, c} from \\"one_module\\";
import d from \\"one_module\\";
import {a, b, c} from \\"one_module\\";
import d from \\"one_module\\";
function __$not$$dash$$source$$dash$$module$$dot$$js__() {
  const x = 1;
  const y = 2;
  const z = 3;
  function square(x) {
    return x * x;
  }
  return pair(square, list(pair(\\"x\\", x), pair(\\"y\\", y), pair(\\"z\\", z)));
}
const ___$not$$dash$$source$$dash$$module$$dot$$js___ = __$not$$dash$$source$$dash$$module$$dot$$js__();
const w = __access_export__(___$not$$dash$$source$$dash$$module$$dot$$js___, \\"default\\");
const x = __access_export__(___$not$$dash$$source$$dash$$module$$dot$$js___, \\"x\\");
const y = __access_export__(___$not$$dash$$source$$dash$$module$$dot$$js___, \\"y\\");
const z = __access_export__(___$not$$dash$$source$$dash$$module$$dot$$js___, \\"z\\");
"
`;

exports[`preprocessFileImports removes all export-related AST nodes 1`] = `
"const x = 42;
let y = 53;
function square(x) {
  return x * x;
}
const id = x => x;
function cube(x) {
  return x * x * x;
}
"
`;

exports[`preprocessFileImports returns a preprocessed program with all imports 1`] = `
"function __$d$$dot$$js__() {
  const addTwo = x => x + 2;
  return pair(null, list(pair(\\"mysteryFunction\\", addTwo)));
}
function __$c$$dot$$js__(___$d$$dot$$js___) {
  const mysteryFunction = __access_export__(___$d$$dot$$js___, \\"mysteryFunction\\");
  const x = mysteryFunction(5);
  function square(x) {
    return x * x;
  }
  return pair(x, list(pair(\\"square\\", square)));
}
function __$b$$dot$$js__(___$c$$dot$$js___) {
  const y = __access_export__(___$c$$dot$$js___, \\"default\\");
  const square = __access_export__(___$c$$dot$$js___, \\"square\\");
  const a = square(y);
  const b = 3;
  return pair(null, list(pair(\\"a\\", a), pair(\\"b\\", b)));
}
const ___$d$$dot$$js___ = __$d$$dot$$js__();
const ___$c$$dot$$js___ = __$c$$dot$$js__(___$d$$dot$$js___);
const ___$b$$dot$$js___ = __$b$$dot$$js__(___$c$$dot$$js___);
const x = __access_export__(___$b$$dot$$js___, \\"a\\");
const y = __access_export__(___$b$$dot$$js___, \\"b\\");
x + y;
"
`;

exports[`preprocessFileImports returns the same AST if the entrypoint file does not contain import/export statements 1`] = `
"function square(x) {
  return x * x;
}
square(5);
"
`;
